cdap-kubernetes/src/test/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironmentTest.java

```diff
@@ -152,4 +195,190 @@ public void testOnNamespaceDeletionWithKubernetesNotFoundError() throws Exceptio
+    String workloadIdentityGCPServiceAccount = "test-service-account@test-project-id.iam.gserviceaccount.com";
+    String workloadIdentityProvider = "https://gkehub.googleapis.com/projects/test-project-id/locations/global/" +
+      "memberships/test-cluster";
+    kubeMasterEnvironment.setNamespaceCreationEnabled();
+    kubeMasterEnvironment.setWorkloadIdentityEnabled();
+    kubeMasterEnvironment.setWorkloadIdentityPool(workloadIdentityPool);
+    kubeMasterEnvironment.setWorkloadIdentityGcpServiceAccountName(workloadIdentityGCPServiceAccount);
+    kubeMasterEnvironment.setWorkloadIdentityProvider(workloadIdentityProvider);
+    when(coreV1Api.readNamespace(eq(KUBE_NAMESPACE), any(), any(), any()))
+      .thenThrow(new ApiException(HttpURLConnection.HTTP_NOT_FOUND, "namespace not found"));
+    when(coreV1Api.readNamespacedConfigMap(any(), any(), any(), any(), any()))
+      .thenThrow(new ApiException(HttpURLConnection.HTTP_INTERNAL_ERROR, "internal error"));
+    when(coreV1Api.createNamespacedConfigMap(any(), any(), any(), any(), any())).thenThrow(new ApiException());
+    kubeMasterEnvironment.onNamespaceCreation(CDAP_NAMESPACE, properties);
+  }
+
+  @Test(expected = ApiException.class)
+  public void testOnNamespaceCreationWithWorkloadIdentityEnabledCreateNamespacePropagatesException() throws Exception {
+    Map<String, String> properties = new HashMap<>();
+    properties.put(KubeMasterEnvironment.NAMESPACE_PROPERTY, KUBE_NAMESPACE);
+    String workloadIdentityPool = "test-workload-pool";
+    String workloadIdentityGCPServiceAccount = "test-service-account@test-project-id.iam.gserviceaccount.com";
+    String workloadIdentityProvider = "https://gkehub.googleapis.com/projects/test-project-id/locations/global/" +
+      "memberships/test-cluster";
+    kubeMasterEnvironment.setNamespaceCreationEnabled();
+    kubeMasterEnvironment.setWorkloadIdentityEnabled();
+    kubeMasterEnvironment.setWorkloadIdentityPool(workloadIdentityPool);
+    kubeMasterEnvironment.setWorkloadIdentityGcpServiceAccountName(workloadIdentityGCPServiceAccount);
+    kubeMasterEnvironment.setWorkloadIdentityProvider(workloadIdentityProvider);
+    when(coreV1Api.readNamespace(eq(KUBE_NAMESPACE), any(), any(), any()))
+      .thenThrow(new ApiException(HttpURLConnection.HTTP_NOT_FOUND, "namespace not found"));
+    when(coreV1Api.readNamespacedConfigMap(eq(KUBE_NAMESPACE), eq("workload-identity-config"), any(), any(), any()))
+      .thenThrow(new ApiException(HttpURLConnection.HTTP_NOT_FOUND, "config map not found"));
+    when(coreV1Api.createNamespacedConfigMap(any(), any(), any(), any(), any())).thenThrow(new ApiException());
+    kubeMasterEnvironment.onNamespaceCreation(CDAP_NAMESPACE, properties);
+  }
+
+  private void assertContainsVolume(V1Pod pod, V1Volume expectedVolume) {
+    for (V1Volume volume : pod.getSpec().getVolumes()) {
+      if (volume.getName().equals(expectedVolume.getName())) {
+        Assert.assertEquals(expectedVolume, volume);
```

Comment:
If the volume as a whole is expected to be equal, why can't you just do `contains` instead of this for loop?

Reply:
I specifically wanted to perform the assert on the volume so that, if the volumes are different, it prints out the difference in the volumes. This makes it a bit easier to debug the tests.

Reply:
If we used `contains`, if a volume with the same name was in the list but did not have the same properties, it would just return "volume not found" unless we dug into the list to explicitly print it out (which would defeat the purpose of using `contains` in the first place).